### PR TITLE
[WIP] Discussion on merging BMXNet 2 contributions

### PR DESCRIPTION
**Pull request by Jopyth**
## Description ##
This PR can help us **start a discussion** on how to integrate the functions required for Binary Neural Networks (BNN) models into mxnet based on our code in BMXNet 2 ( https://github.com/hpi-xnor/BMXNet-v2 ). It should definitely not be merged in its current state.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
We added three functions `det_sign` ([ada4ea1d](https://github.com/hpi-xnor/BMXNet-v2/commit/ada4ea1d4418cfdd6cbc6d0159e1a716cb01cd85)), `round_ste` ([044f81f0](https://github.com/hpi-xnor/BMXNet-v2/commit/044f81f028887b9842070df28b28de394bd07516)) and `contrib.gradcancel` to MXNet (see [src/operator/contrib/gradient_cancel[-inl.h|.cc|.cu]](src/operator/contrib)). - _The `gradcancel` operator is needed for training a BNN only, however it currently is also part of the binary layers._

The rest of the changes was made in the following folders/files:
- Examples are in a submodule in [example/bmxnet-examples](https://github.com/hpi-xnor/BMXNet-v2-examples) - _might reside in a binary model zoo instead_
- Tests are in [tests/binary](tests/binary) - _we run our tests with `pytest`, so a conversion (and possibly integration into the existing test files) of these to `nosetests` should be made_
- Layers are in [python/mxnet/gluon/nn/binary_layers.py](python/mxnet/gluon/nn/binary_layers.py) - _needed for inference and training_
- Converter is in [tools/binary_converter](tools/binary_converter) - _only needed for optimized inference_

See our [Changelog](https://github.com/hpi-xnor/BMXNet-v2/blob/master/CHANGELOG.md) for a (hopefully) complete overview.

## Comments ##
- Since we merged different mxnet releases one after another into our repository, a squash or interactive rebase could help clean the changes up.

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -0,0 +1,71 @@
+stages:
```
Comment (zhreshold) [inline .gitlab-ci.yml]: this file should not go in to master

File: .gitmodules [hunk 2]
```diff
@@ -29,3 +29,6 @@
 [submodule "3rdparty/nvidia_cub"]
 	path = 3rdparty/nvidia_cub
 	url = https://github.com/NVlabs/cub.git
+[submodule "example/bmxnet-examples"]
```
Comment (zhreshold) [inline .gitmodules]: These examples should copy to examples instead of submodule

File: README.md [hunk 3]
```diff
@@ -1,116 +1,136 @@
...
-
-Tianqi Chen, Mu Li, Yutian Li, Min Lin, Naiyan Wang, Minjie Wang, Tianjun Xiao,
-Bing Xu, Chiyuan Zhang, and Zheng Zhang.
-[MXNet: A Flexible and Efficient Machine Learning Library for Heterogeneous Distributed Systems](https://github.com/dmlc/web-data/raw/master/mxnet/paper/mxnet-learningsys.pdf).
-In Neural Information Processing Systems, Workshop on Machine Learning Systems, 2015
-
-History
--------
-MXNet emerged from a collaboration by the authors of [cxxnet](https://github.com/dmlc/cxxnet), [minerva](https://github.com/dmlc/minerva), and [purine2](https://github.com/purine/purine2). The project reflects what we have learned from the past projects. MXNet combines aspects of each of these projects to achieve flexibility, speed, and memory efficiency.
+# BMXNet 2 // Hasso Plattner Institute
```
Comment (zhreshold) [inline README.md]: Leave the README as is, and move this README to examples/bmxnet2
